lib/ABIAnalyses/ABIAnalysis.cpp

```diff
@@ -0,0 +1,299 @@
+  std::map<std::pair<MetaAddress, BasicBlock *>, RegisterStateMap> URVOFC;
+  std::map<std::pair<MetaAddress, BasicBlock *>, RegisterStateMap> RAOFC;
+  std::map<std::pair<MetaAddress, BasicBlock *>, RegisterStateMap> DRVOFC;
+
+  // Per return analysis
+  RegisterStateMap URVOF;
+
+  // Debug methods
+  void dump() const debug_function { dump(dbg, ""); }
+
+  template<typename T>
+  void dump(T &Output, const char *Prefix) const;
+};
+
+// Print the analysis results
+template<typename T>
+void PartialAnalysisResults::dump(T &Output, const char *Prefix) const {
+  Output << Prefix << "UsedArgumentsOfFunction:\n";
+  for (auto &[GV, State] : UAOF) {
+    Output << Prefix << "  " << GV->getName().str() << " = "
+           << model::RegisterState::getName(State).str() << '\n';
+  }
+
+  Output << Prefix << "DeadRegisterArgumentsOfFunction:\n";
+  for (auto &[GV, State] : DRAOF) {
+    Output << Prefix << "  " << GV->getName().str() << " = "
+           << model::RegisterState::getName(State).str() << '\n';
+  }
+
+  Output << Prefix << "UsedReturnValuesOfFunctionCall:\n";
+  for (auto &[Key, StateMap] : URVOFC) {
+    Output << Prefix << "  " << Key.second->getName().str() << '\n';
+    for (auto &[GV, State] : StateMap) {
+      Output << Prefix << "    " << GV->getName().str() << " = "
+             << model::RegisterState::getName(State).str() << '\n';
+    }
+  }
+
+  Output << Prefix << "RegisterArgumentsOfFunctionCall:\n";
+  for (auto &[Key, StateMap] : RAOFC) {
+    Output << Prefix << "  " << Key.second->getName().str() << '\n';
+    for (auto &[GV, State] : StateMap) {
+      Output << Prefix << "    " << GV->getName().str() << " = "
+             << model::RegisterState::getName(State).str() << '\n';
+    }
+  }
+
+  Output << Prefix << "DeadReturnValuesOfFunctionCall:\n";
+  for (auto &[Key, StateMap] : DRVOFC) {
+    Output << Prefix << "  " << Key.second->getName().str() << '\n';
+    for (auto &[GV, State] : StateMap) {
+      Output << Prefix << "    " << GV->getName().str() << " = "
+             << model::RegisterState::getName(State).str() << '\n';
+    }
+  }
+
+  Output << Prefix << "UsedReturnValuesOfFunction:\n";
+  for (auto &[GV, State] : URVOF) {
+    Output << Prefix << "  " << GV->getName().str() << " = "
+           << model::RegisterState::getName(State).str() << '\n';
+  }
+}
+
+static RegisterState combine(RegisterState LH, RegisterState RH) {
+  switch (LH) {
+  case RegisterState::Yes:
+    switch (RH) {
+    case RegisterState::Yes:
+    case RegisterState::YesOrDead:
+    case RegisterState::Maybe:
+      return RegisterState::Yes;
+    case RegisterState::No:
+    case RegisterState::NoOrDead:
+    case RegisterState::Dead:
+    case RegisterState::Contradiction:
+      return RegisterState::Contradiction;
+    case RegisterState::Invalid:
+      revng_abort();
+    }
+    break;
+
+  case RegisterState::YesOrDead:
+    switch (RH) {
+    case RegisterState::Yes:
+    case RegisterState::Maybe:
```

Comment:
`Maybe` intersection `YesOrDead` is `YesOrDead`.
Please review all of them carefully.